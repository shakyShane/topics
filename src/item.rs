use crate::command::Command;
use crate::dependency::DependencyCheck;
use crate::file_exists::FileExistsCheck;
use crate::host::HostEntriesCheck;
use crate::instruction::Instruction;
use crate::topic::Topic;

#[derive(Debug, Clone, serde::Deserialize)]
#[serde(tag = "kind")]
pub enum Item {
    Command(Command),
    FileExistsCheck(FileExistsCheck),
    DependencyCheck(DependencyCheck),
    Instruction(Instruction),
    HostEntriesCheck(HostEntriesCheck),
    Topic(Topic),
}

#[derive(Debug, Clone, serde::Deserialize)]
#[serde(untagged)]
pub enum ItemWrap {
    Named(String),
    Item(Item),
}

impl Item {
    pub fn name(&self) -> String {
        match self {
            Item::Command(cmd) => cmd.name.clone(),
            Item::FileExistsCheck(fec) => fec.name.clone(),
            Item::DependencyCheck(dc) => dc.name.clone(),
            Item::Instruction(inst) => inst.name.clone(),
            Item::HostEntriesCheck(hec) => hec.name.clone(),
            Item::Topic(top) => top.name.clone(),
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_deserialize() -> anyhow::Result<()> {
        let input = r#"
        kind: Topic
        name: run tests
        steps:
            - kind: Instruction
              name: call
              instruction: Call your manager
        "#;
        let t: Topic = serde_yaml::from_str(input)?;
        dbg!(t);
        Ok(())
    }
}
