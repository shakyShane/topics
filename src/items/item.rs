use crate::host::HostEntriesCheck;
use crate::items::{DependencyCheck, TaskGroup};
use crate::items::FileExistsCheck;
use crate::items::Topic;
use crate::items::{Command, Instruction};

#[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
#[serde(tag = "kind")]
pub enum Item {
    Command(Command),
    FileExistsCheck(FileExistsCheck),
    DependencyCheck(DependencyCheck),
    Instruction(Instruction),
    HostEntriesCheck(HostEntriesCheck),
    Topic(Topic),
    TaskGroup(TaskGroup),
}

#[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
#[serde(untagged)]
pub enum ItemWrap {
    Named(String),
    Item(Item),
}

impl Item {
    pub fn name(&self) -> String {
        match self {
            Item::Command(cmd) => cmd.name.clone(),
            Item::FileExistsCheck(fec) => fec.name.clone(),
            Item::DependencyCheck(dc) => dc.name.clone(),
            Item::Instruction(inst) => inst.name.clone(),
            Item::HostEntriesCheck(hec) => hec.name.clone(),
            Item::Topic(top) => top.name.clone(),
            Item::TaskGroup(tg) => tg.name.clone()
        }
    }
}

#[cfg(test)]
mod test {
    use super::*;

    fn test_item() -> Item {
        let input = r#"
        kind: Topic
        name: run tests
        steps:
            - kind: Instruction
              name: call
              instruction: Call your manager
        "#;
        serde_yaml::from_str(input).expect("test doesn't fail")
    }

    #[test]
    fn test_deserialize() -> anyhow::Result<()> {
        let _ = test_item();
        Ok(())
    }

    #[test]
    fn test_serialize() -> anyhow::Result<()> {
        // let t = test_item();
        let item = Item::Topic(Topic::default());
        let as_str = serde_yaml::to_string(&item)?;
        println!("|{}|", as_str);

        Ok(())
    }
}
