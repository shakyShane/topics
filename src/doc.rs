use std::collections::HashMap;
use std::fmt::{Display, Formatter};
use std::path::PathBuf;
use std::str::FromStr;

use crate::items::item::Item;
use crate::{
    context::Context,
    doc_src::DocSource,
    items::{Command, DependencyCheck, Instruction, Topic},
};

#[derive(Debug, Default)]
pub struct Doc {
    pub input_file: PathBuf,
    pub source: DocSource,
    pub topics: HashMap<String, Topic>,
    pub instructions: HashMap<String, Instruction>,
    pub dep_checks: HashMap<String, DependencyCheck>,
    pub commands: HashMap<String, Command>,
    pub errors: Vec<DocError>,
}

pub type DocResult<T, E = DocError> = core::result::Result<T, E>;

impl Doc {
    pub fn from_path_buf(pb: &PathBuf, ctx: &Context) -> DocResult<Self> {
        let doc_src = DocSource::from_path_buf(&pb, ctx)?;
        Self::from_doc_src(&pb, doc_src, &ctx)
    }
    pub fn from_doc_src(pb: &PathBuf, doc_srcs: DocSource, _ctx: &Context) -> DocResult<Self> {
        let mut doc = Doc::default();
        doc.input_file = pb.clone();
        doc.source = doc_srcs;
        lazy_static::lazy_static! {
            static ref RE: regex::Regex = regex::Regex::new("at line (\\d+)").unwrap();
        }
        for src in &doc.source.doc_src_items.items {
            let item: Result<Item, DocError> = serde_yaml::from_str(&src.content).map_err(|e| {
                let mut err = LocationError {
                    input_file_src: doc.source.file_content.clone(),
                    location: Some(Location::Region {
                        line_start: src.line_start + 1,
                        line_end: src.line_end,
                    }),
                    input_file: doc.input_file.clone(),
                    description: e.to_string(),
                };
                if let Some(location) = e.location() {
                    let real_line = location.line() + src.line_start;
                    err.location = Some(Location::LineAndCol {
                        line: real_line,
                        column: location.column(),
                        line_start: src.line_start + 1,
                        line_end: src.line_end,
                    });
                    err.description = RE
                        .replace_all(
                            err.description.as_str(),
                            format!("at line {}", real_line).as_str(),
                        )
                        .to_string()
                }
                DocError::SerdeYamlErr(err)
            });

            match item {
                Err(doc_err) => {
                    doc.errors.push(doc_err);
                }
                Ok(item) => match item {
                    Item::Command(cmd) => {
                        doc.commands.insert(cmd.name.clone(), cmd.clone());
                    }
                    Item::FileExistsCheck(_) => {}
                    Item::DependencyCheck(dc) => {
                        doc.dep_checks.insert(dc.name.clone(), dc.clone());
                    }
                    Item::Instruction(inst) => {
                        doc.instructions.insert(inst.name.clone(), inst.clone());
                    }
                    Item::HostEntriesCheck(_) => {}
                    Item::Topic(t) => {
                        doc.topics.insert(t.name.clone(), t.clone());
                    }
                    Item::TaskGroup(_) => {
                        println!("TaskGroup not implemented yet")
                    }
                },
            };
        }
        Ok(doc)
    }
}

#[derive(Debug, thiserror::Error)]
pub enum DocError {
    #[error(
    "could not read file `{}`\nFull path: {}",
    pb.display(),
    abs.display()
    )]
    PathRead {
        pb: PathBuf,
        abs: PathBuf,
        original: std::io::Error,
    },
    #[error(
        "{}",
        .0
    )]
    SerdeYamlErr(LocationError),
}

#[derive(Debug)]
pub struct LocationError {
    pub location: Option<Location>,
    pub input_file: PathBuf,
    pub input_file_src: String,
    pub description: String,
}

#[derive(Debug)]
pub enum Location {
    LineAndCol {
        line_start: usize,
        line_end: usize,
        line: usize,
        column: usize,
    },
    Region {
        line_start: usize,
        line_end: usize,
    },
    // Unknown,
}

impl Display for LocationError {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let _ = writeln!(f, "");
        if let Some(location) = &self.location {
            match location {
                Location::LineAndCol { line, column, .. } => {
                    let _ = writeln!(f, "    msg: {}", self.description);
                    let _ = writeln!(f, "   file: {}", self.input_file.display());
                    let _ = writeln!(f, "   line: {}", line);
                    let _ = writeln!(f, " column: {}", column);
                }
                Location::Region {
                    line_start,
                    line_end,
                } => {
                    let _ = writeln!(f, "           msg: {}", self.description);
                    let _ = writeln!(f, "          file: {}", self.input_file.display());
                    let _ = writeln!(f, " between lines: {} & {}", line_start, line_end);
                }
            }
        }
        Ok(())
    }
}

impl FromStr for Doc {
    type Err = DocError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let doc_srcs = DocSource::from_str(s)?;
        Doc::from_doc_src(&PathBuf::new(), doc_srcs, &Default::default())
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_single_doc() -> anyhow::Result<()> {
        let input = r#"
kind: Topic
name: Run screen shot tests
deps:
  - global-node
  - global-yarn
steps:
  - github-checkin
"#;
        let doc = Doc::from_str(input)?;
        assert_eq!(doc.source.doc_src_items.items.len(), 1);
        Ok(())
    }

    #[test]
    fn test_multi_doc() -> anyhow::Result<()> {
        let input = r#"
---
kind: Topic
name: Run screen shot tests
deps:
  - global-node
  - global-yarn
steps:
  - github-checkin
---
---
---
kind: Instruction
instruction: help me!
name: help-me-instruction
---
---
"#;
        let doc = Doc::from_str(input)?;
        assert_eq!(doc.source.doc_src_items.items.len(), 2);
        Ok(())
    }

    #[test]
    fn test_errors_single() -> anyhow::Result<()> {
        let pb = PathBuf::from("/input-yaml.yml");
        let input = r#"
kind: Topic
name: Run screen shot tests
deps
"#;
        let srcs = DocSource::from_str(input)?;
        let doc = Doc::from_doc_src(&pb, srcs, &Default::default());
        insta::assert_debug_snapshot!(doc);
        Ok(())
    }

    #[test]
    fn test_errors_multi() -> anyhow::Result<()> {
        let pb = PathBuf::from("/input-yaml.yml");
        let input = r#"---

kind: DependencyCheck
name: global-node
verify: node -v
url: https://www.nodejs.org

---

kind: DependencyCheck
name: global-yarn
verify: yarn -v
url: https://yarn.sh/legacy

---

kind: Topic
name: Run screen shot tests
deps:
  - global-node
  - global-yarn
steps
"#;
        let srcs = DocSource::from_str(input)?;
        let doc = Doc::from_doc_src(&pb, srcs, &Default::default());
        insta::assert_debug_snapshot!(doc?.errors);
        Ok(())
    }
}
