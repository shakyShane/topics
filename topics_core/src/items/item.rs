use crate::host::HostEntriesCheck;
use crate::items::Topic;
use crate::items::{Command, Instruction};
use crate::items::{DependencyCheck, TaskGroup};
use crate::items::{FileExistsCheck, LineMarker};
use std::str::FromStr;

#[derive(Debug, Clone)]
pub enum Item {
    Command(Command),
    FileExistsCheck(FileExistsCheck),
    DependencyCheck(DependencyCheck),
    Instruction(Instruction),
    HostEntriesCheck(HostEntriesCheck),
    Topic(Topic),
    TaskGroup(TaskGroup),
}

#[derive(Debug, Clone)]
pub enum ItemWrap {
    Named(String),
    Item(Item),
}

impl Item {
    pub fn set_name(&mut self, name: &str) {
        match self {
            Item::Command(cmd) => cmd.name = name.into(),
            Item::FileExistsCheck(fec) => fec.name = name.to_string(),
            Item::DependencyCheck(dc) => dc.name = name.to_string(),
            Item::Instruction(inst) => inst.name = name.into(),
            Item::HostEntriesCheck(hec) => hec.name = name.to_string(),
            Item::Topic(top) => top.name = name.into(),
            Item::TaskGroup(tg) => tg.name = name.to_string(),
        };
    }
    pub fn name(&self) -> String {
        match self {
            Item::Command(cmd) => cmd.name.to_string(),
            Item::FileExistsCheck(fec) => fec.name.clone(),
            Item::DependencyCheck(dc) => dc.name.clone(),
            Item::Instruction(inst) => inst.name.to_string(),
            Item::HostEntriesCheck(hec) => hec.name.clone(),
            Item::Topic(top) => top.name.to_string(),
            Item::TaskGroup(tg) => tg.name.clone(),
        }
    }
    pub fn kind_name(&self) -> String {
        match self {
            Item::Command(_) => "Command",
            Item::FileExistsCheck(_) => "File Exists Check",
            Item::DependencyCheck(_) => "Dependency Check",
            Item::Instruction(_) => "Instruction",
            Item::HostEntriesCheck(_) => "Host Entries Check",
            Item::Topic(_) => "Topic",
            Item::TaskGroup(_) => "Task Group",
        }
        .to_string()
    }
    pub fn set_line_start(&mut self, line_start: usize) {
        match self {
            Item::Instruction(inst) => inst.name.set_line_start(line_start),
            Item::Command(cmd) => cmd.name.set_line_start(line_start),
            Item::Topic(topic) => topic.name.set_line_start(line_start),
            _i => todo!("set line start {}", _i.name()),
        }
    }
}

impl FromStr for Item {
    type Err = anyhow::Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "FileExistsCheck" | "fec" => Ok(Item::FileExistsCheck(Default::default())),
            "Topic" | "topic" => Ok(Item::Topic(Default::default())),
            "TaskGroup" | "tg" | "task-group" => Ok(Item::TaskGroup(Default::default())),
            "Command" | "command" | "cmd" => Ok(Item::Command(Default::default())),
            "Instruction" | "inst" | "instruction" => Ok(Item::Instruction(Default::default())),
            "DependencyCheck" | "Dependency Check" | "dep" | "dep-check" => {
                Ok(Item::DependencyCheck(Default::default()))
            }
            _s => Err(anyhow::anyhow!("Not supported yet: {}", _s)),
        }
    }
}
