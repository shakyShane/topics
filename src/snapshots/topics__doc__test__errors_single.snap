---
source: src/doc.rs
expression: doc

---
Ok(
    Doc {
        input_file: "/input-yaml.yml",
        source: DocSource {
            file_content: "\nkind: Topic\nname: Run screen shot tests\ndeps\n",
            doc_src_items: DocSourceItems {
                items: [
                    DocSourceItem {
                        line_start: 0,
                        line_end: 4,
                        content: "\nkind: Topic\nname: Run screen shot tests\ndeps",
                    },
                ],
            },
        },
        topics: {},
        instructions: {},
        dep_checks: {},
        commands: {},
        errors: [
            SerdeYamlErr(
                LocationError {
                    location: Some(
                        LineAndCol {
                            line_start: 1,
                            line_end: 4,
                            line: 4,
                            column: 1,
                        },
                    ),
                    input_file: "/input-yaml.yml",
                    input_file_src: "\nkind: Topic\nname: Run screen shot tests\ndeps\n",
                    description: "while parsing a block mapping, did not find expected key at line 4 column 1",
                },
            ),
        ],
    },
)
